odoo/addons/base/models/res_currency.py

```diff
@@ -60,6 +61,12 @@ def _get_rates(self, company, date):
         currency_rates = dict(self._cr.fetchall())
         return currency_rates
 
+    @api.constrains('active')
+    def _check_company_currency_stays_active(self):
+        for currency in self:
+            if not currency.active and self.env['res.company'].search([('currency_id', '=', currency.id)]):
```

Comment:
search in a for loop in bad juju... :suspect: 

use something like
`companies = self.env['res.company'].search([('currency_id', 'in', self)])
`